File: go/classic/longestPalindrome_5.go
package classic

func longestPalindrome(s string) string {
	memo := make([][]bool, len(s))
	for i := 0; i < len(s); i++ {
		memo[i] = make([]bool, len(s))
	}

	res := ""

	for i := len(s) - 1; i >= 0; i-- {
		for j := i; j < len(s); j++ {
			if i == j || (s[i] == s[j] && (i+1 == j || memo[i+1][j-1])) {
				memo[i][j] = true
			}
			if memo[i][j] && j-i+1 > len(res) {
				res = s[i : j+1]
			}
		}
	}

	return res
}

File: java/src/LongestPalindrome_5.java
public class LongestPalindrome_5 {

    public String longestPalindrome(String s) {
        String res = "";
        boolean[][] dp = new boolean[s.length()][s.length()];
        for (int i = s.length() - 1; i >= 0; i--) {
            for (int j = i; j < s.length(); j++) {
                dp[i][j] = i == j || ((s.charAt(i) == s.charAt(j) && (i + 1 == j || dp[i + 1][j - 1])));
                if (dp[i][j] && j - i + 1 > res.length()) {
                    res = s.substring(i, j + 1);
                }
            }
        }
        return res;
    }
}
